### DISCUSSION TITLE
Correct way to handle authorization

### DISCUSSION THREAD

**Discussion by ftognetto** [vercel/next.js#59815] [App Router]
Hello everyone, I'm new to Next.js and I'm trying to find which is the correct way to handle authorization.
With this I mean deny already logged users but with the wrong role to access some routes.

For example right now I'm working on a project where a user can be an admin or a customer.
I have a feature in which admin can create quotes and customer can accept or deny that quote. But generally speaking I need to find a way to handle the authorization.

1) With middleware I can create different routes for example /admin-quotes and /customer-quotes and handle the authorization in the middleware. I think this is all server side and safe
2) With parallel routing I can create a unique route /quotes and add @admin and @customer slots and handle this in the layout.tsx of the parallel route. I find this very clean but is it safe? 
3) Obviously I can (like the doc says https://nextjs.org/docs/pages/building-your-application/routing/authenticating#authenticating-server-rendered-pages) handle this at page level but my app Is very big and I'm trying to find a generalized way to handle this.

**Comment by ananuness** [comment]
As your page is big, you can go with the first option!! You can leave all this work with middleware and redirect the user to another page if they're accessing a unauthorized page. Here you can see an [example of handling roles using middleware](https://github.com/vercel/next.js/discussions/59884#discussioncomment-7930922) I suggested in this discussion.

**Reply by ftognetto** [reply]
Hi @ananuness thanks for your reply!
Yes I'm actually handling authentication with middleware!
I was wondering for authorization instead, where I have single features and I have to distinguish between a normal and an admin user if parallel routes where a correct pattern to apply for security concern.
Using the middleware for this purpose will lead to many if inside and I'm worried that as the project grow it will became a mess instead managing the authorization at feature level with the parallel route seems very clean to me

**Reply by ananuness** [reply]
Oooh, okay! Now I got it 😅 and yes! Parallel routes in this case would be great and in the docs this is even a [suggestion of use](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes), it's a clean way to not leave it all for middleware and it will make your code look well organized and easier to maintain 🤩
